pkg/upgrade/upgrademanager/manager_external_test.go

```diff
@@ -232,6 +234,139 @@ RETURNING id;`, firstID).Scan(&secondID))
 	require.NoError(t, <-upgrade2Err)
 }
 
+// TestPostJobInfoTableQueryDuplicateJobInfo tests that the
+// PostJobInfoTableQuery returns 1 row for a migration even when the
+// job_info table has 2 payloads for the relevant migration.
+//
+// This condition shouldn't really be possible for the migrations that
+// this query will be used against since it is only possible for jobs
+// that exist during the job_info backfill. By definition, any job we
+// are looking for with this query should be ones started after the
+// query.
+//
+// But, in case we are wrong about that reasoning, we handle it and
+// test it here.
+func TestPostJobInfoTableQueryDuplicateJobInfo(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	targetCV := clusterversion.V23_1StopWritingPayloadAndProgressToSystemJobs + 1
+	targetCVJSON, err := protoreflect.MessageToJSON(&clusterversion.ClusterVersion{Version: clusterversion.ByKey(targetCV)},
+		protoreflect.FmtFlags{EmitDefaults: false})
+	require.NoError(t, err)
+
+	settingsForUpgrade := func() *cluster.Settings {
+		settings := cluster.MakeTestingClusterSettingsWithVersions(
+			clusterversion.TestingBinaryVersion,
+			clusterversion.TestingBinaryMinSupportedVersion,
+			false, // initializeVersion
+		)
+		require.NoError(t, clusterversion.Initialize(ctx,
+			clusterversion.ByKey(clusterversion.BinaryMinSupportedVersionKey), &settings.SV))
+		return settings
+	}
+
+	upgradeStarted := make(chan chan struct{})
+	registryOverrideHook := func(v roachpb.Version) (upgradebase.Upgrade, bool) {
+		if v != clusterversion.ByKey(targetCV) {
+			return nil, false
+		}
+		return upgrade.NewTenantUpgrade("test", v, upgrade.NoPrecondition, func(
+			ctx context.Context, version clusterversion.ClusterVersion, deps upgrade.TenantDeps,
+		) error {
+			canResume := make(chan struct{})
+			upgradeStarted <- canResume
+			select {
+			case <-canResume:
+				return nil
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}), true
+	}
+
+	ts, systemSQLDB, _ := serverutils.StartServer(t, base.TestServerArgs{
+		DefaultTestTenant: base.TestControlsTenantsExplicitly,
+		Settings:          settingsForUpgrade(),
+		Knobs: base.TestingKnobs{
+			JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			Server: &server.TestingKnobs{
+				BinaryVersionOverride:          clusterversion.TestingBinaryMinSupportedVersion,
```

Comment:
should we also be bootstrapping to BinaryMinSupportedVersion so the migrations run from that version until the target version? I'm probably missing something here

Reply:
I think by default we bootstrap to the BinaryMinSupportedVersion in this case:

```
				// If BinaryVersionOverride is set, and our `binaryMinSupportedVersion`
				// is at its default value, we must populate the cluster with initial
				// data from the `binaryMinSupportedVersion`. This cluster will then run
				// the necessary upgrades until `BinaryVersionOverride` before being
				// ready to use in the test.
				if knobs.BinaryVersionOverride != (roachpb.Version{}) {
					if initCfg.binaryMinSupportedVersion.Equal(
						clusterversion.ByKey(clusterversion.BinaryMinSupportedVersionKey)) {
						initialValuesOpts.OverrideKey = clusterversion.BinaryMinSupportedVersionKey
					}
				}
```